src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -0,0 +1,956 @@
+        return instance._ss.str();
+    }
+};
+
+DukContext::DukContext()
+{
+    _context = duk_create_heap_default();
+    if (_context == nullptr)
+    {
+        throw std::runtime_error("Unable to initialise duktape context.");
+    }
+}
+
+DukContext::~DukContext()
+{
+    duk_destroy_heap(_context);
+}
+
+ScriptEngine::ScriptEngine(InteractiveConsole& console, IPlatformEnvironment& env)
+    : _console(console)
+    , _env(env)
+    , _hookEngine(*this)
+{
+}
+
+void ScriptEngine::Initialise()
+{
+    auto ctx = (duk_context*)_context;
+    ScConfiguration::Register(ctx);
+    ScConsole::Register(ctx);
+    ScContext::Register(ctx);
+    ScDate::Register(ctx);
+    ScDisposable::Register(ctx);
+    ScMap::Register(ctx);
+    ScNetwork::Register(ctx);
+    ScObject::Register(ctx);
+    ScPark::Register(ctx);
+    ScPlayer::Register(ctx);
+    ScPlayerGroup::Register(ctx);
+    ScRide::Register(ctx);
+    ScRideObject::Register(ctx);
+    ScTile::Register(ctx);
+    ScTileElement::Register(ctx);
+    ScEntity::Register(ctx);
+    ScPeep::Register(ctx);
+
+    dukglue_register_global(ctx, std::make_shared<ScConsole>(_console), "console");
+    dukglue_register_global(ctx, std::make_shared<ScContext>(_execInfo, _hookEngine), "context");
+    dukglue_register_global(ctx, std::make_shared<ScDate>(), "date");
+    dukglue_register_global(ctx, std::make_shared<ScMap>(ctx), "map");
+    dukglue_register_global(ctx, std::make_shared<ScNetwork>(ctx), "network");
+    dukglue_register_global(ctx, std::make_shared<ScPark>(), "park");
+
+    _initialised = true;
+    _pluginsLoaded = false;
+    _pluginsStarted = false;
+
+    InitSharedStorage();
+}
+
+void ScriptEngine::LoadPlugins()
+{
+    if (!_initialised)
+    {
+        Initialise();
+    }
+    if (_pluginsLoaded)
+    {
+        UnloadPlugins();
+    }
+
+    auto base = _env.GetDirectoryPath(DIRBASE::USER, DIRID::PLUGIN);
+    if (Path::DirectoryExists(base))
+    {
+        auto pattern = Path::Combine(base, "*.js");
+        auto scanner = std::unique_ptr<IFileScanner>(Path::ScanDirectory(pattern, true));
+        while (scanner->Next())
+        {
+            auto path = std::string(scanner->GetPath());
+            if (ShouldLoadScript(path))
+            {
+                LoadPlugin(path);
+            }
+        }
+
+        if (gConfigPlugin.enable_hot_reloading && network_get_mode() == NETWORK_MODE_NONE)
+        {
+            SetupHotReloading();
+        }
+    }
+    _pluginsLoaded = true;
+    _pluginsStarted = false;
+}
+
+void ScriptEngine::LoadPlugin(const std::string& path)
+{
+    auto plugin = std::make_shared<Plugin>(_context, path);
+    LoadPlugin(plugin);
+}
+
+void ScriptEngine::LoadPlugin(std::shared_ptr<Plugin>& plugin)
+{
+    try
+    {
+        ScriptExecutionInfo::PluginScope scope(_execInfo, plugin, false);
+        plugin->Load();
+
+        auto metadata = plugin->GetMetadata();
+        if (metadata.MinApiVersion <= OPENRCT2_PLUGIN_API_VERSION)
```

Comment:
That was not mentioned in the FAQ. How often do we intend to increment this? Each change to exposed functions (either their list or source)?

Reply:
Scripting readme has been updated.